helix-view/src/editor.rs

```diff
@@ -1419,7 +1421,25 @@ impl Editor {
         // via stream.then() ? then push into main future
 
         let path = path.map(|path| path.into());
+        let config = self.config();
         let doc = doc_mut!(self, &doc_id);
+
+        // handle adding newline at the end of the file if there is none
+        let newline = doc.line_ending.as_str();
+        let text = doc.text();
+        let doc_len = text.len_chars();
+        if config.newline_at_eof && get_line_ending(&text.slice(..)).is_none() {
+            let view = view_mut!(self);
+            let old_selection = doc.selection(view.id).clone();
+            let newline = Tendril::from(newline);
+            let selection = Selection::point(doc_len);
+            let transaction = Transaction::insert(text, &selection, newline);
+            doc.set_selection(view.id, selection);
+            doc.apply(&transaction, view.id);
+            doc.append_changes_to_history(view);
+            doc.set_selection(view.id, old_selection);
```

Comment:
```suggestion
            doc.apply(&transaction, view.id);
            doc.append_changes_to_history(view);
```

I have applied the same changes in https://github.com/helix-editor/helix/commit/ab0c01235ae75d612867080db26c067e9c233dcb if that's simpler to use.